VCS microstate with multiple margin tools.

ralesi opened the pull request:
Add a microstate for staging hunks with diff-hl

nixmaniack commented:
This is nice! However it would be good if we can take inputs from this https://github.com/syl20bnr/spacemacs/issues/3585
Like staging, committing and pushing right from the current buffer instead of visiting magit. Would that be possible to add to this microstate? Just a heads up, when I looked at diff-hl last time there was no support for staging as it was based on vc IIRC.

syl20bnr commented:
`gh` is now reserved for `github` key bindings.
I'm for get rid of the `evil-leader/set-key` bindings and move the micro-state to `gdH` (`gdh` is already taken for diff-head), so the first action is something like `gdHn` then we can press only `n` etc...

ralesi commented:
@syl20bnr, what about making gD for hunk then?  gdH is a little too much.  We can make gd for diff views, (like magit-diff- and magit-ediff) and gD for hunks within buffer.

syl20bnr commented:
Seems more reasonable indeed :-)

ralesi commented:
Good, now what about evil commands?  Does anyone even use `ex` commands anymore?  I used to love being able to `:Gwrite` with `fugitive.vim` to add the file I am working on to the index.

ralesi commented:
I just completely redid this commit.  Now have options to use `diff-hl`, `git-gutter`, or `git-gutter+` as margin tools, each with their own vcs micro-state, using `SPC g.`

nixmaniack commented:
Going to try this out! Sounds promising. It would be better if we document the pros and cons/differences of each tool. As I expected, diff-hl doesn't have the ability to stage hunks, so the default now seems to be git-gutter+.

nixmaniack commented:
The doc string goes out of way in default gitter+, Can we stick to some limit or at least screen width? If possible, show only commonly used options in default state, and provide detailed doc toggle on `?`.

nixmaniack commented on layers/+source-control/version-control/packages.el:
```diff
@@ -32,8 +38,188 @@
+    :config
+    (set-face-foreground 'git-gutter+-modified "black")
+    (set-face-foreground 'git-gutter+-added    "black")
+    (set-face-foreground 'git-gutter+-deleted  "black")
```
Do you think if it's good idea to set colors explicitly? They might not blend with theme colors. I don't know the right solution but it somehow inheriting from current theme. Anyhow, seems like a personal preference, if it can't be handled, you may ignore.

nixmaniack commented on layers/+source-control/version-control/packages.el:
```diff
@@ -32,8 +38,188 @@
+      ("w" magit-stage-file)
+      ("u" magit-unstage-file)
+      ("h" git-gutter+-show-hunk-inline-at-point)
+      ("t" global-git-gutter+-mode)
```
This should exit micro state as well? Or move this toggle to `SPC t v`?
Also, you might want to enable this mode when entering this microstate else most of the hunk navigation commands won't work.

ralesi commented:
Yes colors should be likely a config,  or use the default diff faces like diff-hl uses. Outside of the staging portion,  there arent very clear differences that I can see.  Git-gutter+ has more features,  but I think only git-gutter and diff-hl work with vcs other than git.  The doc toggle was another thought,  I also need to work on standardizing the keys used for each tool,  or define a single micro-state that is able to pull from whichever tool is used.

ralesi commented:
I am also concerned with using magit as part of the state.  Would make more sence to just use vc,  but magit is so much nicer...

ralesi commented:
Just added improved doc string and separated out all the tool commands so we can have one complete micro-state.

nixmaniack commented:
Looks much cleaner. :+1:  May I suggest a binding for previous hunk on `p` as well to be consistent with other micro states?

ralesi commented:
Was thinking that as well,  any more magit bindings?

StreakyCobra commented on layers/+source-control/git/README.org:
```diff
@@ -192,19 +194,19 @@ will discard the commit message.
-| ~u~         | undo           |
-| ~y~         | insert         |
-| ~!~         | execute        |
+| ~c~ or ~p~      | pick           |
```
Don't forget to align the table before you final version :-)

nixmaniack commented:
This key set looks complete to me now. If we need anything later we might add it then.

StreakyCobra commented on layers/+source-control/version-control/README.org:
```diff
@@ -25,11 +28,31 @@ To use this contribution add it to your =~/.spacemacs=
+
+VCS Hunk Micro-state
+
+| Key Binding | Description                   |
```
Horizontal line missing in this table, it's needed for `Key Binding | Description` to be shown as a header.

StreakyCobra commented on layers/+source-control/version-control/README.org:
```diff
@@ -25,11 +28,31 @@ To use this contribution add it to your =~/.spacemacs=
+|-------------+---------------------------|
+| ~SPC g .~   | vcs hunk micro-state      |
+
+VCS Hunk Micro-state
```
This will look better as a subsection, maybe… not sure :-)

ralesi commented:
@StreakyCobra,  @robbyoconnor, @nixmaniack filled out the layer a bit more, added customization and proper spacemacs-toggles.  Also updated the readme with apparent differences I can see betwene the packages.  Let me know if you have any more suggestions.

nixmaniack commented on layers/+source-control/git/packages.el:
```diff
@@ -112,13 +112,21 @@
+        "gw" 'magit-stage-file
+        "gu" 'magit-unstage-file
+        "gE" 'magit-ediff-compare
+        "ge" 'magit-ediff-show-working-tree)
```
Changes here seem unrelated to vcs micro state. I assume you don't want to change the spacemacs' default bindings in this PR. Probably a mistake?

ralesi replied:
This was part of the original diff overhaul as party of the original commit,  I can take out and add separately.

ralesi commented:
I have rebased to HEAD and squashed all commits.  I think this is feature complete at this point.  Let me know if you have any more suggestions.

robbyoconnor commented:
@ralesi, you sir are amazing. :tada:

CestDiego commented:
This is so cool! :) thank you @ralesi

nixmaniack commented:
:+1: Awesome!

ralesi commented:
Should I rebase this to current HEAD?

TheBB commented:
That would be great.

syl20bnr commented:
Thank you ! :100: 
This is a wonderful commit, great work!
I made minor fixes like conversion of the newline format of new files and moving `T D` to `T C-d` (which is the convention for the global version of a toggle), see commits following the pull of this commit.
Cherry-picked into develop branch, you can safely delete your branch.

robbyoconnor commented:
:+1:
On 11/29/2015 11:14 PM, Sylvain Benner wrote:
> Thank you ! :100:
> This is a wonderful commit, great work!
> I made minor fixes like conversion of the newline format of new files
> and moving |T D| to |T C-d| (which is the convention for the global
> version of a toggle), see commits following the pull of this commit.
> Cherry-picked into develop branch, you can safely delete your branch.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/syl20bnr/spacemacs/pull/3632#issuecomment-160512221>.
>
